src/term/string.rs

```diff
@@ -0,0 +1,399 @@
+    /// `Iterator`. However, rather than iterating through grapheme clusters
+    /// directly, its `Item` is a [`SearchEvent`].
+    pub struct SearchIter<'a> {
+        haystack: &'a str,
+        needle: &'a str,
+        cursor: GraphemeCursor,
+        current_split_start: usize,
+        last_boundary: usize,
+    }
+
+    impl<'a> Iterator for SearchIter<'a> {
+        type Item = SearchEvent<'a>;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            let unwrap_explanation = "
+                None of the GraphemeIncomplete errors are possible here:
+                    - PreContext and PrevChunk only happen if chunk_start is nonzero.
+                    - NextChunk only happens if the chunk is smaller than the cursor's len parameter
+                      but we pass `haystack` and `hackstack.len()`` respectively.
+                    - InvalidOffset can't happen because we check that `haystack` contains `needle`
+                      in the range (last_boundary, last_boundary + needle.len())
+            ";
+
+            loop {
+                let nxt = self
+                    .cursor
+                    .next_boundary(self.haystack, 0)
+                    .expect(unwrap_explanation);
+                match nxt {
+                    Some(next_boundary) => {
+                        // To check whether a match has occured, we'll first check whether
+                        // the slice of `haystack` starting at the `last_boundary` begins
+                        // with `needle`. If it does, we *also* need to check that this
+                        // instance of `needle` doesn't end in the middle of a cluster.
+                        // This is to avoid the situtaion where we have a string like
+                        // `"a🧑‍🤝‍🧑"` and we get a match by searching for `"a🧑"`.
+                        let does_match_intersect_cluster = || {
+                            let mut tmp_cursor = self.cursor.clone();
+                            tmp_cursor.set_cursor(self.last_boundary + self.needle.len());
+                            !tmp_cursor
+                                .is_boundary(self.haystack, 0)
+                                .expect(unwrap_explanation)
+                        };
+
+                        if self.haystack[self.last_boundary..].starts_with(self.needle)
+                            && !does_match_intersect_cluster()
+                        {
+                            // Yay, we found a match! We start by grabbing the slice of
+                            // the string we traversed while looking for the match.
+                            let since_last_match =
+                                &self.haystack[self.current_split_start..self.last_boundary];
+                            // Now we move the cursor to after the match, to avoid needlessly
+                            // traversing that part of the string again.
```

Comment:
If you're looking for, say, "aa" in "aaaa", this will produce only two matches, one at offset 0 and the other at offset 2, won't it?

On reflection, this doesn't matter in our case, since all our string primitives either look for the first match or look to do a search-and-replace, where you want this skipping behaviour explicitly. But maybe it's worth explicitly documenting that?